src/service.js

```diff
@@ -0,0 +1,879 @@
+const debug = Debug('api');
+
+/* In production, log authorizations so they are included in papertrail regardless of
+ * DEBUG settings; otherwise, log with debug
+ */
+let authLog = (...args) => console.log(...args);
+if (process.env.NODE_ENV !== 'production') {
+  authLog = Debug('api.authz');
+}
+
+/**
+ * A ping method, added automatically to every service
+ */
+const ping = {
+  method:   'get',
+  route:    '/ping',
+  name:     'ping',
+  stability:  'stable',
+  title:    'Ping Server',
+  description: [
+    'Respond without doing anything.',
+    'This endpoint is used to check that the service is up.',
+  ].join('\n'),
+  handler: function(req, res) {
+    res.status(200).json({
+      alive:    true,
+      uptime:   process.uptime(),
+    });
+  },
+};
+
+/** Return [route, params] from route */
+const _cleanRouteAndParams = (route) => {
+  // Find parameters for entry
+  const params = [];
+  // Note: express uses the NPM module path-to-regexp for parsing routes
+  // when modifying this to support more complicated routes it can be
+  // beneficial lookup the source of this module:
+  // https://github.com/component/path-to-regexp/blob/0.1.x/index.js
+  route = route.replace(/\/:(\w+)(\(.*?\))?\??/g, (match, param) => {
+    params.push(param);
+    return '/<' + param + '>';
+  });
+  return [route, params];
+};
+
+/**
+ * A service represents an instance of an API at a specific rootUrl, ready to
+ * provide an Express router, references, etc.  It is constructed by API.setup().
+ */
+class Service {
+  constructor(options) {
+    assert(!options.authBaseUrl, 'authBaseUrl option is no longer allowed');
+    assert(!options.baseUrl, 'baseUrl option is no longer allowed');
+    assert(options.api, 'api option is required');
+    assert(options.rootUrl, 'rootUrl option is required');
+    assert(!options.referencePrefix, 'referencePrefix is now deprecated!');
+
+    options = _.defaults({}, options, {
+      inputLimit:           '10mb',
+      allowedCORSOrigin:    '*',
+      context:              {},
+      nonceManager:         nonceManager(),
+      referenceBucket:      'references.taskcluster.net',
+      signatureValidator:   createRemoteSignatureValidator({
+        rootUrl: options.rootUrl,
+      }),
+    });
+    this.api = options.api;
+
+    // validate context
+    this.api.context.forEach(function(property) {
+      assert(options.context[property] !== undefined,
+        'Context must have declared property: \'' + property + '\'');
+    });
+
+    Object.keys(options.context).forEach(property => {
+      assert(this.api.context.indexOf(property) !== -1,
+        `Context has unexpected property: ${property}`);
+    });
+
+    // make `entries` specific to this rootUrl
+    this.entries = options.api.entries.map((entry) => {
+      entry = _.clone(entry);
+
+      // fully-qualify schema references
+      if (entry.input) {
+        assert(!entry.input.startsWith('http'), 'entry.input should be a filename, not a url');
```

Comment:
Were we doing this in any places before?

Reply:
Oh, I see this down below now. Really interesting that we ever did things like that.